client/hudwidget.cpp

```diff
@@ -1731,3 +1770,135 @@ void hud_battle_log::showEvent(QShowEvent *event)
   m_timer.restart();
   setVisible(true);
 }
+
+/**
+   This predicate is called by cycle_units and friends to filter
+   relevant units from the iterated units.
+
+   current_unit is the currently selected unit.
+   candidate is the candidate checked for relevance.
+ */
+static bool cycle_units_predicate(const struct unit *current_unit,
+                                  const struct unit *candidate)
+{
+  fc_assert_ret_val(current_unit != nullptr && candidate != nullptr, false);
+
+  if (candidate->utype != current_unit->utype) {
+    return false;
+  }
+
+  if (ACTIVITY_IDLE != candidate->activity
+      && ACTIVITY_SENTRY != candidate->activity) {
+    return false;
+  }
+
+  if (!can_unit_do_activity(candidate, ACTIVITY_IDLE)) {
+    return false;
+  }
+
```

Comment:
Should we exclude sentried units in boats?

Reply:
The checks in `cycle_units_predicate` are based on the `find nearest` algorithm used by the units report: 

https://github.com/longturn/freeciv21/blob/df33c26984b67cea7f86c2bfd99f1dc6c97ac421/client/views/view_units.cpp#L553-L558

I think it makes sense not to introduce a changed behaviour for this functionality in order to keep it consistent.

Reply:
Two things here:
* Can we avoid duplicating the logic?
* I'm not convinced by the `can_unit_do_activity` check, we can select units without cancelling their orders. The [original commit](https://github.com/freeciv/freeciv/commit/e839b2595f7d5234338f091c49ac4e825090c233) doesn't provide much information. We should probably open a new issue to look into this further.

Reply:
* I factored out the logic into a function in `controls.cpp`. Looking at the other functions in this file, I figured it is fitting.
* The `can_unit_do_activity` check was also in the original cycle units code in the old units report. I guess the new code was based on this and inherited the check.

https://github.com/longturn/freeciv21/blob/3003835fa85260bafdd48f4895c0da86b0cdf149/client/unitreport.cpp#L362-L367